File: algorithms/713. Subarray Product Less Than K/solution.py
class Solution(object):
    def numSubarrayProductLessThanK(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        ret = end = 0
        p = 1
        for start in xrange(0, len(nums)):
            if end < start:
                end = start
            while end < len(nums) and p * nums[end] < k:
                p *= nums[end]
                end += 1
            ret += end - start
            p /= nums[start]
        return ret

# 64 / 64 test cases passed.
# Status: Accepted
# Runtime: 349 ms

File: algorithms/713. Subarray Product Less Than K/solution.cc
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int result = 0, end = 0, p = 1;
        for ( int start = 0; start < nums.size(); ++start ) {
            end = max(end, start);
            while ( end < nums.size() && p * nums[end] < k ) {
                p *= nums[end];
                ++end;
            }
            result += end - start;
            p /= nums[start];
        }
        return result;
    }
};

// 64 / 64 test cases passed.
// Status: Accepted
// Runtime: 93 ms
